src/cljs/main/broadfcui/common/markdown.cljs

```diff
@@ -32,29 +32,37 @@
 
 (react/defc MarkdownEditor
   {:get-trimmed-text
-   (fn [{:keys [state]}]
+   (fn [{:keys [state this]}]
+     (assert (not (this :-controlled?)) "Trying to read text from a controlled component")
      (:text @state))
    :get-initial-state
-   (fn [{:keys [props]}]
-     {:mode :edit
-      :text (or (:initial-text props) "")})
+   (fn [{:keys [props this]}]
+     (merge
+      {:mode :edit}
+      (when-not (this :-controlled?)
+        {:text (or (:initial-text props) "")})))
    :render
-   (fn [{:keys [props state]}]
-     (let [{:keys [mode text]} @state
+   (fn [{:keys [props state this]}]
+     (let [{:keys [mode]} @state
            tab (fn [mode-key label]
                  (let [selected? (= mode-key mode)]
                    [:span {:style {:border style/standard-line :padding "3px 8px" :cursor "pointer"
                                    :color (when selected? "white")
                                    :backgroundColor ((if selected? :button-primary :background-light) style/colors)}
                            :onClick #(swap! state assoc :mode mode-key)}
                     label]))
+           controlled? (this :-controlled?)
+           text (if controlled? (:value props) (:text @state))
            markdown-view [MarkdownView {:text text}]
-           text-area (style/create-text-area {:value (:text @state)
-                                              :onChange #(swap! state assoc :text (-> % .-target .-value))
+           text-area (style/create-text-area {:value text
+                                              :onChange #(let [new-value (-> % .-target .-value)]
+                                                           (if controlled?
+                                                             ((:on-change props) new-value)
+                                                             (swap! state assoc :text new-value)))
                                               :style {:width "100%" :height "100%" :minHeight 200
                                                       :resize (if (= mode :edit) "vertical" "none")}})]
        [:div {}
-        [:div {}
+        [:div {:style {:marginTop "0.5rem"}}
          (tab :edit "Edit")
          (tab :preview "Preview")
          (tab :side-by-side "Side-by-side")]
```

Comment:
I don't know if it was introduced here or earlier, but the buttons seem to be overlapped by the view in edit and side-by-side (try making the preview side wide enough that it starts before the buttons end)

Reply:
Looks like it's always been like that, but I fixed it.